src/storage/src/upsert/types.rs

```diff
@@ -623,16 +670,71 @@ where
     where
         G: IntoIterator<Item = UpsertKey>,
         R: IntoIterator<Item = &'r mut UpsertValueAndSize<O>>;
+
+    /// For each key in `merges` writes a 'merge operand' to the backend. The backend stores these
+    /// merge operands and periodically calls the `snapshot_merge_function` to merge them into
+    /// any existing value for each key. The backend will merge the merge operands in the order
+    /// they are provided, and the merge function will always be run for a given key when a `get`
+    /// operation is performed on that key, or when the backend decides to run the merge based
+    /// on its own internal logic.
+    /// This allows avoiding the read-modify-write method of updating many values to
+    /// improve performance.
+    ///
+    /// `MergeStats` **must** be populated correctly, according to these semantics:
+    /// - `written_merge_operands` must record the number of merge operands written to the backend.
+    /// - `size_written` must record the total size of values written to the backend.
+    /// Note that the size of the post-merge values are not known, so this is the size of the values
+    /// written to the backend as merge operands.
+    async fn multi_merge<P>(&mut self, merges: P) -> Result<MergeStats, anyhow::Error>
+    where
+        P: IntoIterator<Item = (UpsertKey, StateValue<O>)>;
+}
+
+/// A function that merges a set of updates for a key into the existing value for the key, expected
+/// to only be used during the snapshotting-phase of an upsert operator. This is called by the
+/// backend implementation when it has accumulated a set of updates for a key, and needs to merge
+/// them into the existing value for the key.
+///
+/// The function should return the new value for the key, or None if the key should be deleted.
```

Comment:
Just found out that this is an incorrect description for the merge operator rust semantics. It turns out returning `None` indicates to rocksdb that the 'merge operator' failed and will cause a `Corruption` error message.

The rust docs for the merge operator aren't descriptive what the return-type means https://docs.rs/rocksdb/latest/rocksdb/merge_operator/index.html unfortunately so I figured this out by testing.

So it doesn't appear that you can 'delete' a key from the result of this merge function, so I'll always need to return a `Some` value here and rework this logic.

Reply:
I think you should just return a value, and `Some`-wrap it in the `mz-rocksdb` crate when we actually wrap the function so it can be passed to the api.

It returns `Some` because `rust-rocksdb` was lazy and naively translated the `*buffer` of the c api into an `Option<buffer>`, despite the documentation requiring non-`NULL`